packages/esm-patient-test-results-app/src/panel-view/usePanelData.tsx

```diff
@@ -0,0 +1,194 @@
+      url += `&_count=100`;
+      if (pageIndex) {
+        url += `&_getpagesoffset=${pageIndex * 10}`;
+      }
+      return url;
+    },
+    [patientUuid],
+  );
+  const { data, error, size, setSize } = useSWRInfinite<FetchResponse<FhirResponse<FHIRObservationResource>>, Error>(
+    getUrl,
+    openmrsFetch,
+    {
+      revalidateIfStale: false,
+      revalidateOnFocus: false,
+    },
+  );
+
+  useEffect(() => {
+    // Infinitely fetching all the data
+    if (data && data?.length === size && data?.[data.length - 1]?.data?.link?.some((x) => x.relation === 'next')) {
+      setSize(size + 1);
+    }
+  }, [size, setSize, data]);
+
+  const results = useMemo(() => {
+    const observations: Array<FHIRObservationResource> = data
+      ? []
+          .concat(...data?.map((resp) => resp.data?.entry?.map((e) => e.resource) ?? []))
+          .sort((obs1, obs2) => Date.parse(obs2.effectiveDateTime) - Date.parse(obs1.effectiveDateTime))
+      : null;
+    return {
+      observations,
+      isLoading: !data && !error,
+      conceptUuids: observations ? [...new Set(observations.map((obs) => getConceptUuid(obs)))] : null,
+    };
+  }, [data, error]);
+
+  return results;
+}
+
+function useConcepts(conceptUuids: Array<string>) {
+  const getUrl = (index) => {
+    if (conceptUuids && index < conceptUuids.length) {
+      return `/ws/rest/v1/concept/${conceptUuids[index]}?v=full`;
+    }
+    return null;
+  };
+  const { data, error } = useSWRInfinite<FetchResponse<Concept>>(getUrl, openmrsFetch, {
+    initialSize: conceptUuids?.length ?? 1,
+    revalidateIfStale: false,
+    revalidateOnFocus: false,
+  });
+
+  const results = useMemo(() => {
+    const concepts: Array<Concept> = data ? [].concat(data?.map((resp) => resp.data)) : null;
+    return {
+      concepts: concepts
+        ? concepts.filter((c) => c.conceptClass.display === 'Test' || c.conceptClass.display === 'LabSet')
+        : null,
+      // If there are no observations, hence no concept UUIDS, then it should return isLoading as false
+      isLoading: conceptUuids?.length === 0 ? false : !data && !error,
+    };
+  }, [data, error, conceptUuids]);
+
+  return results;
+}
+
+export default function usePanelData() {
+  const { observations: fhirObservations, conceptUuids, isLoading: isLoadingObservations } = useObservations();
+  const { isLoading: isLoadingConcepts, concepts } = useConcepts(conceptUuids);
+
+  const conceptData: Record<string, ConceptMeta> = useMemo(
+    () =>
+      concepts
+        ? Object.fromEntries(
+            concepts?.map((concept) => [
+              concept.uuid,
+              {
+                display: concept.display,
+                ...extractMetaInformation(concept),
+              },
+            ]),
+          )
+        : {},
+    [concepts],
+  );
+
+  const observations: Array<ObsRecord> = fhirObservations?.map((observation) => {
+    const conceptUuid = getConceptUuid(observation);
+    let value: number;
```

Comment:
IMO we should we initialize this to *something* even if its `undefined` just to make the default behavior clear.